cli/azd/test/recording/poll_discard.go

```diff
@@ -0,0 +1,363 @@
+package recording
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"gopkg.in/dnaeon/go-vcr.v3/cassette"
+)
+
+// httpPollDiscarder discards awaiting-done polling interactions from the cassette.
+// As a result of this, the cassette will only contain the final result of the polling operation.
+// When playing back interactions from the cassette, this results in fast-forwarding behavior.
+//
+// The type of polling protocols httpPollDiscarder detects includes:
+// - Azure specific async polling protocols (implementation closely matches ones supported by azure-sdk-for-go)
+// - Idiomatic HTTP async polling protocols (Location header, 201,202 status codes)
+type httpPollDiscarder struct {
+	pollInProgress poller
+}
+
+func (d *httpPollDiscarder) BeforeSave(i *cassette.Interaction) error {
+	if d.pollInProgress != nil && d.pollInProgress.Done(i) {
+		// Polling is done, reset pollInProgress
+		d.pollInProgress = nil
+	}
+
+	// Discard awaiting-done polling interactions
+	if d.pollInProgress != nil {
+		i.DiscardOnSave = true
+		return nil
+	}
+
+	// Check if the interaction is a polling request.
+	// If so, set the current poll in progress.
+	if op, err := NewPoller(i); op != nil {
+		if err != nil {
+			return err
+		}
+		d.pollInProgress = op
+
+		// Shorten Retry-After
+		if i.Response.Headers.Get("Retry-After") != "" {
+			i.Response.Headers.Set("Retry-After", "0")
+		}
+	}
+
+	return nil
+}
+
+func NewPoller(i *cassette.Interaction) (poller, error) {
+	// accepted status codes for polling operations
+	if c := i.Response.Code; c != http.StatusOK &&
+		c != http.StatusAccepted &&
+		c != http.StatusCreated &&
+		c != http.StatusNoContent {
+		return nil, nil
+	}
+
+	// order here matches the order in azruntime.Poller
+	if isAsyncPoll(i) {
+		return newAsyncPoll(i)
+	}
+
+	if isOpPoll(i) {
+		return newOpPoll(i)
+	}
+
+	if isLocationPoll(i) {
+		return newLocationPoll(i)
+	}
+
+	if isBodyPoll(i) {
+		return newBodyPoll(i)
+	}
+
+	return nil, nil
+}
```

Comment:
I'd like to understand more about this polling special use case and why it requires special handling.  No matter what the Azure polling implementation does behind the scenes they are all just standard HTTP requests that return a response, right?

Reply:
Yes, it's all done through HTTP. But HTTP doesn't define a standard for client polling. Thus, early days of Azure decided to use custom headers like "Azure-AsyncOperation" or "Operation-Location", along with custom headers and bodies that represent the status of the resource.

Generally speaking, a Location header and 201/202 status codes indicate operation being in-progress.

Reply:
Yes, I understand how the polling is implemented.  I'm more concerned why we need to care and have special handling logic for recorder tests for this.

Reply:
Quoting the documentation again:

```
// httpPollDiscarder discards awaiting-done polling interactions from the cassette.
// As a result of this, the cassette will only contain the final result of the polling operation.
// When playing back interactions from the cassette, this results in fast-forwarding behavior.
```

We don't "have" to implement any of these to discard interactions, but the playback ends up rerunning in the same duration as the recording. Does that make sense?